### PR TITLE
Fix resource usage tracking and remove stale mapper

### PR DESCRIPTION
**Pull request by rhc54**
[Fix resource usage tracking for map/bind operations](https://github.com/openpmix/prrte/commit/0acc016860bab1099879ca09dc3f12015e29f82a)

Tracking solely at the slot level doesn't adequately protect against
overlapping CPU assignments and other more complex mapping requests.
Modify the resource usage tracking to operate at the CPU level and
combine the mapping/binding operation into a single pass.

Ranking must still be done as a second pass, but restrict the options
to simplify implementation and avoid confusion. Update the help output
to reflect the changes.

Allow the DVM to also support "do-not-launch" directives for testing
purposes, and to accept simulated node/topologies.

Fix a few other minor problems along the way.

Signed-off-by: Ralph Castain <rhc@pmix.org>

[Remove the mindist mapper - no maintainer](https://github.com/openpmix/prrte/commit/368bed34f3ba51137be48cdded0f7fc8263ed57f)

Signed-off-by: Ralph Castain <rhc@pmix.org>

### CODE REVIEW

**Comment by rhc54** [comment]
@jjhursey Your CI contains an error in one of its cmd lines:
```text
prterun --map-by ppr:5:node --host 192.168.210.115:5,192.168.160.63:5 hostname
```
cannot possibly succeed. By definition, the `--host` option will only assign one slot to each of the specified hosts. Thus, this cmd line is required to error out with the "not enough slots" error, which is precisely what it does.

The cmd line would work for the case where the allocation was given by a resource manager. It just cannot work in environments where the allocation is given by a hostfile or is only given via the `--host` cmd line option.

I'll correct this in the `pmix-tests` repo, but wanted to explain why it had to be changed.

**Comment by rhc54** [comment]
Funny enough, it actually worked in one environment - which leads me to suspect you must have specified a default hostfile in there somewhere. I don't rule out that it might be working/failing randomly due to some other factor. However, it _should_ fail every time as-written.

**Comment by rhc54** [comment]
Ignore the noise - I saw that the cmd line does indeed specify the `:5` slot number as it should. Something else must be going on.

**Comment by rhc54** [comment]
@jjhursey Afraid I am going to need your help with the darned PGI case again. I simply cannot replicate this behavior. It appears that the node pointer in the `PMIX_LIST_FOREACH_SAFE` is not getting advanced for some reason. Can you try to get some reason for this behavior?

**Comment by rhc54** [comment]
Update: I tried modifying the CI script to isolate the test and then adding debug output. Best I could find is that the "foreach" list macro in src/mca/rmaps/ppr/rmaps_ppr.c is in fact operating correctly. However, all the nodes on the list report the same hostname when running in the PGI environment.

I was unable to get any further diagnostics from it, so I can't tell if something strange is in the environment (making all the nodes resolve to the same hostname) or if something odd is happening inside PRRTE. Finally had to give up - this will require someone with access to the VM to debug it.

I reset the CI scripts and removed the debug I had added to the PR. I may have to do a little more cleanup to ensure I "reverting" things cleanly, but otherwise this should be good-to-go.

**Comment by rhc54** [comment]
@jjhursey I believe the `indirect-multi` tests are in fact running correctly, but that the signature has changed due to the revised default ranking algorithms. Could you please double-check to confirm, and if that is correct, then update the signatures? I have found that I cannot replicate things closely enough here to regenerate them for your CI setup.

**Comment by jjhursey** [comment]
bot:ibm:retest

I can take a look at the PGI issue. The new CI environment does set
```
PRTE_MCA_prte_default_hostfile=/opt/mpi/etc/hostfile
```
So there is a default hostfile (trying to enumate a properly scheduled environment). Then use the `--host` option for the few tests that need to restrict down from there.

**Comment by rhc54** [comment]
Thanks! I just pushed a debug change in for the dmodex test, so expect some extra verbage from the OOB. If it gets in your way, I can revert it and work the dmodex problem later.

**Comment by rhc54** [comment]
I have removed that OOB debug - sorry for the noise.

File: src/mca/rmaps/base/rmaps_base_support_fns.c [hunk 1]
```diff
@@ -664,9 +700,225 @@ prte_node_t *prte_rmaps_base_get_starting_point(pmix_list_t *node_list, prte_job
...
-    PRTE_OUTPUT_VERBOSE((5, prte_rmaps_base_framework.framework_output, "%s Starting at node %s",
+    PRTE_OUTPUT_VERBOSE((5, prte_rmaps_base_framework.framework_output,
+                         "%s Starting at node %s",
                          PRTE_NAME_PRINT(PRTE_PROC_MY_NAME),
                          ((prte_node_t *) cur_node_item)->name));
 
-    return (prte_node_t *) cur_node_item;
+    /* put this node at the front of the list */
+    pmix_list_remove_item(node_list, cur_node_item);
+    pmix_list_prepend(node_list, cur_node_item);
```
Comment (jjhursey) [inline src/mca/rmaps/base/rmaps_base_support_fns.c:710]: While debugging the PGI issue I saw the allocated_nodes list become corrupted after calling the `prte_rmaps_base_get_starting_point` function at the bottom of the `prte_rmaps_base_get_target_nodes` function . These two lines seemed to be the cause.

This fixed the PGI issue, but I'm a bit afraid that it is masking a different issue.
```suggestion
    if (pmix_list_get_first(node_list) != cur_node_item) {
        pmix_list_remove_item(node_list, cur_node_item);
        pmix_list_prepend(node_list, cur_node_item);
    }
```

Can you give me a scenario where you would need to execute this section of code to move the node to the front of the list?

**Comment by jjhursey** [comment]
Actually the error above seems to be the problem with the debug example in the other test cases. The `prterun` is just not mapping so the test is failing.

**Comment by rhc54** [comment]
> Actually the error above seems to be the problem with the debug example in the other test cases. The `prterun` is just not mapping so the test is failing.

Not sure what you mean by this statement - are you saying `prterun` isn't mapping due to the corruption you found? Or are you saying it isn't mapping in general?

**Comment by rhc54** [comment]
Oh? Weird - I'll have to take a look at that one.

**Comment by rhc54** [comment]
I think I found this last problem - working on it now. Not sure what to do about the list corruption just yet. Probably needs more investigation.

File: src/mca/rmaps/base/rmaps_base_support_fns.c (hunk 1 above)
Comment (rhc54) [reply src/mca/rmaps/base/rmaps_base_support_fns.c:710]: I dug thru the old PRRTE list functions and compared them to the PMIx functions we now use. There were only two differences:

1. PRRTE uses atomics to modify the refcount and PMIx does not. I don't think this is the issue as we only typically access lists from inside events. Still, something to keep in mind.
2. PMIx was missing a couple of type casts. These were right in the "remove_first" and "remove_last" inline functions, which means they _might_ have had an impact here. I added the type casts in https://github.com/openpmix/openpmix/pull/2642.

Let's see if that makes a difference - if not, then I'll add your fix.

**Comment by rhc54** [comment]
Hmmm...well, GNU ran fine except for this last debugger test:
```text
01:07:00 Starting testcase pid 76982: './indirect-multi --num-nodes 3 --hostfile hostfile_5_slots prterun --hostfile hostfile_5_slots --np 12 ./hello '
01:07:03 Verify stdout/stderr for testcase indirect-multi
10,15c10,12
< daemon-0            : [<host>:PID<?>:@NS<1>@2:0:PID<?>] Local proctable received for nspace '<host>:PID<?>:@NS<2>' has 5 entries
< daemon-0            : Proctable[0], namespace <host>:PID<?>:@NS<2> rank 0 exec hello
< daemon-0            : Proctable[1], namespace <host>:PID<?>:@NS<2> rank 1 exec hello
< daemon-0            : Proctable[2], namespace <host>:PID<?>:@NS<2> rank 2 exec hello
< daemon-0            : Proctable[3], namespace <host>:PID<?>:@NS<2> rank 3 exec hello
< daemon-0            : Proctable[4], namespace <host>:PID<?>:@NS<2> rank 4 exec hello
---
> daemon-0            : [<host>:PID<?>:@NS<1>@2:0:PID<?>] Local proctable received for nspace '<host>:PID<?>:@NS<2>' has 2 entries
> daemon-0            : Proctable[0], namespace <host>:PID<?>:@NS<2> rank 10 exec hello
> daemon-0            : Proctable[1], namespace <host>:PID<?>:@NS<2> rank 11 exec hello
32,36c29,33
< daemon-1            : Proctable[0], namespace <host>:PID<?>:@NS<2> rank 5 exec hello
< daemon-1            : Proctable[1], namespace <host>:PID<?>:@NS<2> rank 6 exec hello
< daemon-1            : Proctable[2], namespace <host>:PID<?>:@NS<2> rank 7 exec hello
< daemon-1            : Proctable[3], namespace <host>:PID<?>:@NS<2> rank 8 exec hello
< daemon-1            : Proctable[4], namespace <host>:PID<?>:@NS<2> rank 9 exec hello
---
> daemon-1            : Proctable[0], namespace <host>:PID<?>:@NS<2> rank 0 exec hello
> daemon-1            : Proctable[1], namespace <host>:PID<?>:@NS<2> rank 1 exec hello
> daemon-1            : Proctable[2], namespace <host>:PID<?>:@NS<2> rank 2 exec hello
> daemon-1            : Proctable[3], namespace <host>:PID<?>:@NS<2> rank 3 exec hello
> daemon-1            : Proctable[4], namespace <host>:PID<?>:@NS<2> rank 4 exec hello
51,53c48,53
< daemon-2            : [<host>:PID<?>:@NS<1>@2:2:PID<?>] Local proctable received for nspace '<host>:PID<?>:@NS<2>' has 2 entries
< daemon-2            : Proctable[0], namespace <host>:PID<?>:@NS<2> rank 10 exec hello
< daemon-2            : Proctable[1], namespace <host>:PID<?>:@NS<2> rank 11 exec hello
---
> daemon-2            : [<host>:PID<?>:@NS<1>@2:2:PID<?>] Local proctable received for nspace '<host>:PID<?>:@NS<2>' has 5 entries
> daemon-2            : Proctable[0], namespace <host>:PID<?>:@NS<2> rank 5 exec hello
> daemon-2            : Proctable[1], namespace <host>:PID<?>:@NS<2> rank 6 exec hello
> daemon-2            : Proctable[2], namespace <host>:PID<?>:@NS<2> rank 7 exec hello
> daemon-2            : Proctable[3], namespace <host>:PID<?>:@NS<2> rank 8 exec hello
> daemon-2            : Proctable[4], namespace <host>:PID<?>:@NS<2> rank 9 exec hello
```
I'm not sure I understand the output here. Based on the cmd line, daemon 0 should host ranks 0-4, daemon 1 should host ranks 5-9, and daemon 2 should host ranks 10-11. Is that what happened? It looks instead like we wound up starting on daemon 1 and then mapping around - is that correct?

I can try to locally reproduce.

On the PGI test case, it simply stops at `run.sh` for `prun-wrapper` without any output. I'll try adding your patch.

**Comment by rhc54** [comment]
@jjhursey Still seeing a few warnings from the MPIR shim:
```text
mpirshim.c: In function 'register_launcher_complete_handler':
mpirshim.c:1135:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
mpirshim.c: In function 'register_launcher_ready_handler':
mpirshim.c:1184:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
mpirshim.c: In function 'register_launcher_terminate_handler':
mpirshim.c:1235:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
mpirshim.c: In function 'register_application_terminate_handler':
mpirshim.c:1286:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
  CCLD     libmpirshim.la
  CCLD     mpirc
  CC       libmpirshimtest_la-mpirshim.lo
mpirshim.c: In function 'register_launcher_complete_handler':
mpirshim.c:1135:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
mpirshim.c: In function 'register_launcher_ready_handler':
mpirshim.c:1184:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
mpirshim.c: In function 'register_launcher_terminate_handler':
mpirshim.c:1235:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
mpirshim.c: In function 'register_application_terminate_handler':
mpirshim.c:1286:26: warning: variable 'rc' set but not used [-Wunused-but-set-variable]
     pmix_status_t event, rc;
                          ^~
  CCLD     libmpirshimtest.la
make[1]: Leaving directory '/opt/ci/support/pmix-tests/prrte/mpir-shim/mpir-to-pmix-guide/src'
Making all in test
make[1]: Entering directory '/opt/ci/support/pmix-tests/prrte/mpir-shim/mpir-to-pmix-guide/test'
  CC       mpirshim_test-mpirshim_test.o
mpirshim_test.c:148:6: warning: no previous prototype for 'MPIR_Breakpoint_hook' [-Wmissing-prototypes]
 void MPIR_Breakpoint_hook(void)
      ^~~~~~~~~~~~~~~~~~~~
```

**Comment by rhc54** [comment]
So what I'm seeing is that the `ppr`, `seq`, and `rank-file` mappers are working just fine (the latter two are tested locally and not in CI). The round-robin mappers also seem to be working okay, but are starting from the wrong node for some reason (at least in the `indirect-multi` case). I'll have to explore and see why - need to reproduce the spawn cmd that `indirect-multi` is generating to see if I can make it fail.

Specifically, it is this cmd line that is the only one failing:
```shell
./indirect-multi --num-nodes 3 --hostfile hostfile_5_slots prterun --hostfile hostfile_5_slots --np 12 ./hello 
```

**Comment by rhc54** [comment]
@jjhursey I finally fixed it, but I'm disturbed by the problem/fix. Your patch didn't quite resolve the problems with that last `indirect-multi` command. I had to revise it a bit as we were still getting the wrong node at the start of the list.

I did some digging around to see if the `prepend` list function is used elsewhere. Obviously, we use `remove first` in a lot of places, so I doubt that is causing the problem. Prepend is used less often, but it is used in the code base. However, it isn't clear if those code paths are actually ever traversed - it is possible therefore that `pmix_list_prepend` has been bad for quite some time and just doesn't get used.

What bothers me is that I didn't touch any of that code path involving the bookmark. I only modified the mapping algorithms themselves. Assembling the list of nodes and traversing them - I didn't change that at all. So why did this now break?

Leaves me a little uneasy that the corruption is actually occurring elsewhere and we are only seeing it when we get to that spot in the procedure. I just can't nail down where it might be happening. Any further guidance would be appreciated.

Thanks for all the help!

**Comment by jjhursey** [comment]
Yeah things are running clean for me too now with your branch. I'm not sure why the list operations would cause an issue - I reviewed their implementations and it seemed ok. `valgrind` on `valgrind prterun --hostfile hostfile_4_slots --np 12 ./hello` didn't show anything unusual so 🤷 We'll just have to see if PGI trips over something in the future.

**Pull request merged**